### DISCUSSION TITLE
Audio experiments on iOS

### DISCUSSION THREAD

**Discussion by maxblomberg** [jspsych/jsPsych#956] [Q&A] [answered]
Hello everybody,

first of all, i found jspsych to be an incredible useful tool and i am currently using it to design a PASAT-like experiment (i.e. listening to numbers and clicking on their images) that we want to run online so people could participate from various devices (computer/phone/tablet).

I've set up an example on cognition.run: https://ucc5bdlzs4.cognition.run/ and you can find the source code below. I am using jspsych 6.1 image-button-response and audio-button-response plugins. 

It works on all systems I have tested so far (windows, macOS, android). However, on iOS-devices (iPhones/iPad), it doesn't. After a while of showing a non-progressing loading bar, an "The experiment failed to load" error-message appears. I also tried it on chrome on my iPhone, but nothing changed.
This might be an audio-specific issue since experiments with only images work.

Currently the media files are uploaded to formr and accessible to everybody. I have also tried a version where I access the media files from my google drive, with the results remaining the same.

Any ideas on what might be the problem? 

Thank you very much in advance!

/* create timeline */
var timeline = [];

    /* define stimuli */
    var pic18 = "https://formr.org/assets/tmp/admin/VlzBP47-NDyLXdyzlxdUKi9_GLyZfbUFxs-I6RQAXlHA.png?v1596451185"
    var pic15 = "https://formr.org/assets/tmp/admin/fzYLU1j0MfIvakEkhjra3nj7lkpAmcfwLud2xuY91hGP.png?v1596451185"
    var pic9 = "https://formr.org/assets/tmp/admin/pNx1XJh1JPyzwCc5OtNDb7mAQIXTOr0l0C3B2UgeWdkR.png?v1596451185"
    var sound15 = "https://formr.org/assets/tmp/admin/wYDkuyHmL7DoJ_CId5qa4EV_LSd3c5PfKOB2xs8qwV7c.mp3?v1596622899"
    var sound9 = "https://formr.org/assets/tmp/admin/n8kOZI1A3hiUcaFuqzhx6GZLhJF1kcVkCQdPqB4m4_8I.mp3?v1596700311"
    var sound18 = "https://formr.org/assets/tmp/admin/LlIgN9aahfYzLjPhGZEBiVo2c1IhSvM9WOKxWDIgTmrf.mp3?v1596700311"

      /* interaction check */
      timeline.push({
      type: 'image-button-response',
      stimulus: pic18,
      choices: ["Yes", "No"],
      prompt: "<p>Do you see a picture of a number?</p>"
      });

      timeline.push({
                type: 'audio-button-response',
                stimulus: sound15,
                choices: [pic18, pic15, pic9],
                prompt: "<p>Which number did you hear?</p>",
                button_html: '<img src="%choice%" />'
      });

      timeline.push({
                type: 'audio-button-response',
                stimulus: sound9,
                choices: [pic18, pic15, pic9],
                prompt: "<p>Which number did you hear?</p>",
                button_html: '<img src="%choice%" />'
      });

      timeline.push({
                type: 'audio-button-response',
                stimulus: sound18,
                choices: [pic18, pic15, pic9],
                prompt: "<p>Which number did you hear?</p>",
                button_html: '<img src="%choice%" />'
      });

    /* start the experiment */
    jsPsych.init({
      timeline: timeline,
      use_webaudio: false,
         })

**Comment by jodeleeuw** [comment, marked as answer]
Hi @maxblomberg 

Have you tried uploading the files to cognition.run? I suspect there is some kind of CORS issue on iOS.

**Reply by maxblomberg** [reply]
Thanks for your fast response, Josh. I tried, but it did not change. I found out that it might have something to do with [iOS disabling autoplay](https://stackoverflow.com/questions/31776548/why-cant-javascript-play-audio-files-on-iphone-safari). However, since my programming skills are limited I still did not manage to create a workaround with the user-interaction enabling playback.

**Reply by jodeleeuw** [reply]
I unfortunately don't have an iOS device to test this on. Because you are having issues getting past the loading stage of the experiment I'm thinking that the autoplay problem is probably not the likely culprit, but I'm not confident about that.

Have you tried setting `use_webaudio: true`?

**Reply by maxblomberg** [reply]
Oh yeah, actually it worked! Never thought of this because all threads on CORS issues suggest setting it to false. However, this means the experiment will not work on non-iOS devices. Think I will look up on how to change the use_webaudio setting based on the OS of the user. Thanks for your help!

**Reply by jodeleeuw** [reply]
It should still work on non-iOS devices. 

You'll just run into issues if you are trying to run the experiment *locally* rather than on a server. But if you are using cognition.run I don't expect you'll have a problem.

**Reply by maxblomberg** [reply]
You're right. For reasons I don't understand setting  `use_webaudio: true` only works when hosting the experiment on cognition.run, but not on formr (where I basically just upload all the jspsych-code to run the experiment), and only when accessing the audio files from the uploads on cognition.run, not with the url from e.g. google-drive. 
Anyways, that's a working solution. Thanks again!

**Comment by ThalesBuzan** [comment]
Hello everyone!

I know that this discussion is pretty much resolved, but I'd like to add a doubt regarding running experiments on iOS. 

At first, I had the issue of not loading the audios at all in iOS devices. To solve that, I added the text below at the beginning of the code:

// preload audios
    var preload_audios = {
        type: 'preload_audios',
        auto_preload: true
    };

This seemed to work, the audios loaded and the experiment moved on, but there was no sound at all playing... I can't seem to solve this problem.

I have to basically identical experiments running through Cognition, the difference being the language (one is in Brazilian Portuguese and the other one is in English). Both of them can be found below.

On the BP version, I could add use_webaudio: true but that did nothing to solve the problem. On the English version, the experiment failed, running only if I used use_webaudio: false .

I am using this platform for the first time and I am programming these experiments for the first time as well. Any ideas on this issue?

// preload audios
    var preload_audios = {
        type: 'preload_audios',
        auto_preload: true
    };
// create timeline
	var timeline = [];
  // survey

	// create form
	var form = {
		type: "survey-html-form",
		preamble: "<strong>Antes de começarmos, gostarímos de saber mais sobre você!</strong>",
		html:"<p>Qual é o seu nome? <input name='name' type='text' placeholder='Ex.: Roberto Silva' required/></p>" +
 
		  "<p>Onde você nasceu? <input name='country' type='text' placeholder='País, estado, cidade' required/></p>" +

			"<p>Onde você reside? <input name='country' type='text' placeholder='País, estado, cidade' required/></p>" +
			
			"<p>Quantos anos você tem? <input  name='age' placeholder= 'Escolha o intervalo' required/ list='ages'><datalist id='ages'><option value='18-25'><option value='26-35'><option value='36-45'><option value='46-55'><option value='56+'></datalist></p>" +

			"<p>Você tem algum conhecimento da língua inglesa? <input name='language' type='text' placeholder='Sim/Não' required/></p>",
		button_label: "OK",
		data: {type: "form"}
	};

	// put form in timeline
	timeline.push(form);
	
		// create likert scale
	var scale = ["Muito baixa", "Baixa", "Média", "Boa", "Muito boa"];

	// create stimuli
	var stimuli = [
		{prompt:"Como você avalia sua habilidade de compreensão oral em inglês?",name:1},
	];
	
	// create trial
	var trial = {
		type: "survey-likert",
		questions: [
			{prompt: jsPsych.timelineVariable("prompt"), name: jsPsych.timelineVariable("name"), labels: scale, required: true}
		],
		button_label: "OK"
	};

	// create procedure
	var procedure = {
		timeline: [trial],
		timeline_variables: stimuli,
		randomize_order: true
	};
	// put practice_procedure in timeline
	timeline.push(procedure);
	
	// create instructions
	var instructions = {
		type: 'html-button-response',
		stimulus:	'<p><strong>INSTRUÇÕES</strong></p></br>' +
					'<p>Você ouvirá alguns áudios. Esses áudios são frases do inglês, mas para os propósitos deste teste, todas as palavras foram removidads - <strong>você ouvirá somente a melodia da frase.</strong></p></br>' +
					'<p>Ao final de cada áudio, há uma pergunta a qual você deve responder se o que acabou de ouvir é uma "PERGUNTA" ou "AFIRMAÇÃO".</p>' +
					//'<p>Pressione a <strong>tecla P</strong> com a mão direita para palavras existentes.</p>' +
					'<p>Lembre-se de <strong>usar fones de ouvido</strong> durante o teste.</p>' +
					'<p>Pressione <strong>OK</strong> para começar.</p>',
		post_trial_gap: 500,
		choices: ['OK']
	};
	// put instructions in timeline
	timeline.push(instructions);

	// create fixation
	var fixation = {
		type: 'html-button-response',
		stimulus: '+',
		choices: jsPsych.NO_KEYS,
		trial_duration: 1000
	};

	// create stimuli
	var stimuli = [
		{stimulus: 'AU_1_Ligia_-_10Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_2_Igor_-_2D3_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_3_Ligia_-_11Q3_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_4_Dalila_-_5D1_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_5_Vitor_-_9Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_6_Dani_-_10Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_7_Luiza_-_1D1_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_8_Fernanda_-_11Q1_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_9_Vitor_-_4D3_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_10_Paula_-_10Q3_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_11_Larissa_-_7Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_12_Dani_-_1D2_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_13_Luiza_-_10Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_14_Lydia_-_8Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_15_Jake_-_2D2_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_16_Ligia_-_12Q3_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_17_Lydia_-_3D1_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_18_Paula_-_9Q3_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_19_Zenia_-_4D3_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_20_Lydia_-_7Q3_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_21_Jake_-_11Q1_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_22_Lydia_-_5D3_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_23_Larissa_-_9Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_24_Pedro_-_8Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_25_Jake_-_6D1_band.mp3', data: {correct_response: 'AFIRMAÇÃO'}},
		{stimulus: 'AU_26_Zenia_-_7Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		{stimulus: 'AU_27_Vitor_-_11Q2_band.mp3', data: {correct_response: 'PERGUNTA'}},
		
	];

	// create test
	var test = {
		type: 'audio-button-response',
		stimulus: jsPsych.timelineVariable('stimulus'),
		choices: ['PERGUNTA','AFIRMAÇÃO'],
		post_trial_gap: 700,
		trial_duration: 20000,

		data: jsPsych.timelineVariable('data'),
		on_finish: function(data){
			data.correct = data.response == data.correct_response;
		}
	};

	// create procedure
		var procedure = {
		timeline: [fixation, test],
		timeline_variables: stimuli
	};
	// put procedure in timeline
	timeline.push(procedure);

	// create end_experiment
	var end = {
		type: 'html-button-response',
		stimulus: '<p>Este é o final do teste.</p>' +
				'<p>Agradecemos a participação!</p>' +
				'<p>Precione <strong>OK</strong> finalizar.</p>',
		trial_duration: 5000,
		choices: ['OK']
	};
	// put end in timeline
	timeline.push(end);

	// run experiment
	jsPsych.init({
		timeline: timeline,
		preload_audio: ["AU_1_Lígia_-_10Q2_band.mp3", "AU_2_Igor_-_2D3_band.mp3", "AU_3_Lígia_-_11Q3_band.mp3", "AU_4_Dalila_-_5D1_band.mp3", "AU_5_Vitor_-_9Q2_band.mp3", "AU_6_Dani_-_10Q2_band.mp3", "AU_7_Luiza_-_1D1_band.mp3", "AU_8_Fernanda_-_11Q1_band.mp3", "AU_9_Vitor_-_4D3_band.mp3", 'AU_10_Paula_-_10Q3_band.mp3', 'AU_11_Larissa_-_7Q2_band.mp3', 'AU_12_Dani_-_1D2_band.mp3', 'AU_13_Luiza_-_10Q2_band.mp3', 'AU_14_Lydia_-_8Q2_band.mp3', 'AU_15_Jake_-_2D2_band.mp3', 'AU_16_Lígia_-_12Q3_band.mp3', 'AU_17_Lydia_-_3D1_band.mp3', 'AU_18_Paula_-_9Q3_band.mp3', 'AU_19_Zenia_-_4D3_band.mp3', 'AU_20_Lydia_-_7Q3_band.mp3', 'AU_21_Jake_-_11Q1_band.mp3','AU_22_Lydia_-_5D3_band.mp3', 'AU_23_Larissa_-_9Q2_band.mp3','AU_24_Pedro_-_8Q2_band.mp3', 'AU_25_Jake_-_6D1_band.mp3','AU_26_Zenia_-_7Q2_band.mp3', 'AU_27_Vitor_-_11Q2_band.mp3'],
		use_webaudio: true,
	});
____________________________________________________________________________

// preload audios
    var preload_audios = {
        type: 'preload_audios',
        auto_preload: true
    };
// create timeline
	var timeline = [];
  // survey

	// create form
	var form = {
		type: "survey-html-form",
		preamble: "<strong>Before we start, we'd like to know more about you!</strong>",
		html:"<p>What is your name? <input name='name' type='text' placeholder='Ex.: Peter Parker' required/></p>" +
 
		  "<p>Where are you from? <input name='country' type='text' placeholder='Country, state, city' required/></p>" +

			"<p>Where do you live in? <input name='country' type='text' placeholder='Country, state, city' required/></p>" +
			
			"<p>How old are you? <input  name='age' placeholder= 'Choose the interval' required/ list='ages'><datalist id='ages'><option value='18-25'><option value='26-35'><option value='36-45'><option value='46-55'><option value='56+'></datalist></p>" +

			"<p>Do you speak any language other than English? <input name='language' type='text' placeholder='Yes/No. Language(s).' required/></p>",
		button_label: "OK",
		data: {type: "form"}
	};
	// put form in timeline
	timeline.push(form);
	
	// create instructions
	var instructions = {
		type: 'html-button-response',
		stimulus:	'<p><strong>INSTRUCTIONS</strong></p></br>' +
					'<p>You will listen to several audios. These audios are English sentences, but for the purposes of this test all words have been removed - <strong>you will only hear the melody of the sentence.</strong></p></br>' +
					'<p>For each audio there is one question in which you must answer if what you just heard is a "QUESTION" or a "STATEMENT".</p>' +
					//'<p>Pressione a <strong>tecla P</strong> com a mão direita para palavras existentes.</p>' +
					'<p>Remember to <strong>use headphones/earbuds</strong> during the test.</p>' +
					'<p>Press <strong>OK</strong> to start.</p>',
		post_trial_gap: 500,
		choices: ['OK']
	};
	// put instructions in timeline
	timeline.push(instructions);

	// create fixation
	var fixation = {
		type: 'html-button-response',
		stimulus: '+',
		choices: jsPsych.NO_KEYS,
		trial_duration: 1000
	};

	// create stimuli
	var stimuli = [
		{stimulus: 'AU_1_Ligia_-_10Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_2_Igor_-_2D3_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_3_Ligia_-_11Q3_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_4_Dalila_-_5D1_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_5_Vitor_-_9Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_6_Dani_-_10Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_7_Luiza_-_1D1_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_8_Fernanda_-_11Q1_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_9_Vitor_-_4D3_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_10_Paula_-_10Q3_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_11_Larissa_-_7Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_12_Dani_-_1D2_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_13_Luiza_-_10Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_14_Lydia_-_8Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_15_Jake_-_2D2_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_16_Ligia_-_12Q3_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_17_Lydia_-_3D1_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_18_Paula_-_9Q3_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_19_Zenia_-_4D3_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_20_Lydia_-_7Q3_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_21_Jake_-_11Q1_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_22_Lydia_-_5D3_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_23_Larissa_-_9Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_24_Pedro_-_8Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_25_Jake_-_6D1_band.mp3', data: {correct_response: 'STATEMENT'}},
		{stimulus: 'AU_26_Zenia_-_7Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		{stimulus: 'AU_27_Vitor_-_11Q2_band.mp3', data: {correct_response: 'QUESTION'}},
		
	];

	// create test
	var test = {
		type: 'audio-button-response',
		stimulus: jsPsych.timelineVariable('stimulus'),
		choices: ['QUESTION', 'STATEMENT'],
		post_trial_gap: 700,
		trial_duration: 20000,

		data: jsPsych.timelineVariable('data'),
		on_finish: function(data){
			data.correct = data.response == data.correct_response;
		}
	};

	// create procedure
		var procedure = {
		timeline: [fixation, test],
		timeline_variables: stimuli
	};
	// put procedure in timeline
	timeline.push(procedure);

	// create end_experiment
	var end = {
		type: 'html-button-response',
		stimulus: '<p>End of the test.</p>' +
				'<p>Thanks for participating!</p>' +
				'<p>Press <strong>SPACE</strong> to end.</p>',
		trial_duration: 5000,
		choices: ['OK']
	};
	// put end in timeline
	timeline.push(end);

	// run experiment
	jsPsych.init({
		timeline: timeline,
		preload_audio: ["AU_1_Lígia_-_10Q2_band.mp3", "AU_2_Igor_-_2D3_band.mp3", "AU_3_Lígia_-_11Q3_band.mp3", "AU_4_Dalila_-_5D1_band.mp3", "AU_5_Vitor_-_9Q2_band.mp3", "AU_6_Dani_-_10Q2_band.mp3", "AU_7_Luiza_-_1D1_band.mp3", "AU_8_Fernanda_-_11Q1_band.mp3", "AU_9_Vitor_-_4D3_band.mp3", 'AU_10_Paula_-_10Q3_band.mp3', 'AU_11_Larissa_-_7Q2_band.mp3', 'AU_12_Dani_-_1D2_band.mp3', 'AU_13_Luiza_-_10Q2_band.mp3', 'AU_14_Lydia_-_8Q2_band.mp3', 'AU_15_Jake_-_2D2_band.mp3', 'AU_16_Lígia_-_12Q3_band.mp3', 'AU_17_Lydia_-_3D1_band.mp3', 'AU_18_Paula_-_9Q3_band.mp3', 'AU_19_Zenia_-_4D3_band.mp3', 'AU_20_Lydia_-_7Q3_band.mp3', 'AU_21_Jake_-_11Q1_band.mp3','AU_22_Lydia_-_5D3_band.mp3', 'AU_23_Larissa_-_9Q2_band.mp3','AU_24_Pedro_-_8Q2_band.mp3', 'AU_25_Jake_-_6D1_band.mp3','AU_26_Zenia_-_7Q2_band.mp3', 'AU_27_Vitor_-_11Q2_band.mp3'],
		use_webaudio: false,
	});

**Reply by maxblomberg** [reply]
Although I am also not so familiar with javascript, two things here look kind of odd to me:

1) The code at the beginning
`var preload_audios = {
type: 'preload_audios',
auto_preload: true
};
`
seems unnecessary to me or at least is not responsible for preloading the audiofiles. The type `preload_audios` you use for that variable is not a valid jsPsych plugin. See here: https://www.jspsych.org/plugins/list-of-plugins/. You could consider using the [jspsych-preload plugin](https://www.jspsych.org/plugins/jspsych-preload/). 

2) In the same fashion, using `preload_audio:` in the jsPsych.init() method is not a valid setting for this method: https://www.jspsych.org/core_library/jspsych-core/#jspsychinit. To my mind, it therefore should not have any effect.

Did you maybe add other js scripts to your experiment where you defined this plugin type and init()-setting?
